### PR TITLE
Raise warning for 3D images in denoise_bilateral

### LINKED ISSUE #1840: denoise_bilateral works separately on channels, but no warning raised for 3-D images
The implementation of ``restoration.denoise_bilateral`` is 2-D only, with the last dimension considered to be for channels. However, this is not documented in the docstring, and no warning is raised if a 3-D image (with last dimension > 3) is passed. 

I used ``denoise_bilateral`` in the past on 3-D images thinking that the implementation was 3-D. We could consider making the code 3-D or nD in the future, but for now fixing the docstring + warning looks like a reasonable goal for 0.12.

### PR DESCRIPTION
**Pull request by OrkoHunter**
Closes #1840; Completes #1842 

@emmanuelle I couldn't find any tests which need to be warped because of this warning. Please suggest changes accordingly.

### CODE REVIEW

**Comment by stefanv** [comment]
@blink1073 @jni What is the recommended way to handle checks for color images?

**Comment by JDWarner** [comment]
We've hashed this out a number of times, with no good answer to the problem - other than depend on the user to properly use the function, or provide us with the information. Magic will always be incorrect for some shallow volumes; e.g., this solution breaks for RGBA images with an alpha channel. It just can't be done automatically.

My personal preferred method is `multichannel=False` as a kwarg. We could expose this kwarg even in functions like this one, which would give us sufficient info to guarantee the correct code path.

**Comment by stefanv** [comment]
That's an interesting notion: ``multichannel=False`` which can't be set to True :)

**Comment by blink1073** [comment]
@stefanv, by best guess: https://github.com/scikit-image/scikit-image/blob/master/skimage/io/_io.py#L66

**Comment by JDWarner** [comment]
@stefanv Well, it can be for a 2D input - denoting a color image. The main benefit is that such a kwarg would allow us to be completely unambiguous at present, and allow a consistent API to be built out in the future with 3D or nD support.

**Comment by stefanv** [comment]
I am happy with either the short term or longer term solution here.

**Comment by OrkoHunter** [comment]
Even with grayscale image, the situation is being handled magically over [here](https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/_denoise_cy.pyx#L56). With the new solution, in case `multichannel is False`, we expect an image with only two dimensions otherwise it's either a 3D image or needs `multichannel=True`. In the case where `multichannel is True`, the function can take either `RGB` or `RGBA` i.e. `image.shape[2] in (3, 4)`. 

@JDWarner Is it similar to what you thought as the solution?

**Comment by JDWarner** [comment]
Yes, that's the sort of solution I had in mind.

That line of code is not being handled by magic - the algorithm is written expecting a third channel, so 2d grayscale images have an additional singleton dimension appended internally. This lets us use the exact same code path.

File: skimage/restoration/_denoise.py [hunk 1]
```diff
@@ -64,6 +68,20 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
     >>> noisy = np.clip(noisy, 0, 1)
     >>> denoised = denoise_bilateral(noisy, sigma_range=0.05, sigma_spatial=15)
     """
+    if multichannel:
+        if image.shape[2] not in (3, 4):
+            msg = "denoise_bilateral is expected to work correctly for " \
```
Comment (stefanv) [inline skimage/restoration/_denoise.py]: How about we change the message to:

``Input image must be grayscale, RGB, or RGBA, but has shape {0}``

File: skimage/restoration/_denoise.py [hunk 2]
```diff
@@ -64,6 +68,20 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
...
+                  "2D grayscale or RGB image; image shape {0} looks like " \
+                  "a 3D image."
+            warnings.warn(msg.format(image.shape))
+    else:
+        if image.ndim > 2:
+            msg = "denoise_bilateral is expected to work correctly for " \
```
Comment (stefanv) [inline skimage/restoration/_denoise.py]: Here we should raise an error, since we cannot handle ``multichannel=True``.

**Comment by OrkoHunter** [comment]
@stefanv Updated.

**Comment by OrkoHunter** [comment]
The tests should pass now.

File: skimage/restoration/_denoise.py [hunk 3]
```diff
@@ -45,6 +46,9 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
     cval : string
         Used in conjunction with mode 'constant', the value outside
         the image boundaries.
+    multichannel : bool, default False
```
Comment (stefanv) [inline skimage/restoration/_denoise.py]: Remove mention of default here (already visible in signature).

File: skimage/restoration/_denoise.py [hunk 4]
```diff
@@ -62,8 +66,21 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
...
     >>> noisy = astro + 0.6 * astro.std() * np.random.random(astro.shape)
     >>> noisy = np.clip(noisy, 0, 1)
-    >>> denoised = denoise_bilateral(noisy, sigma_range=0.05, sigma_spatial=15)
+    >>> denoised = denoise_bilateral(noisy, sigma_range=0.05, \
+                                     sigma_spatial=15, multichannel=True)
     """
+    if multichannel:
+        if image.shape[2] not in (3, 4):
+            msg = "Input image must be grayscale, RGB, or RGBA; but has " \
+                  "a shape {0}."
```
Comment (stefanv) [inline skimage/restoration/_denoise.py]: a shape -> shape

File: skimage/restoration/_denoise.py [hunk 5]
```diff
@@ -62,8 +66,21 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
...
+            warnings.warn(msg.format(image.shape))
+    else:
+        if image.ndim > 2:
+            msg = "Input image must be grayscale, RGB, or RGBA; but has " \
+                  "a shape {0}."
```
Comment (stefanv) [inline skimage/restoration/_denoise.py]: same here

**Comment by OrkoHunter** [comment]
@stefanv Last 3 comments addressed.

File: skimage/restoration/_denoise.py [hunk 6]
```diff
@@ -62,8 +66,21 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
     >>> astro = astro[220:300, 220:320]
     >>> noisy = astro + 0.6 * astro.std() * np.random.random(astro.shape)
     >>> noisy = np.clip(noisy, 0, 1)
-    >>> denoised = denoise_bilateral(noisy, sigma_range=0.05, sigma_spatial=15)
+    >>> denoised = denoise_bilateral(noisy, sigma_range=0.05, \
```
Comment (stefanv) [inline skimage/restoration/_denoise.py]: Remove backslash

File: skimage/restoration/_denoise.py [hunk 7]
```diff
@@ -62,8 +66,21 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
...
+                                     sigma_spatial=15, multichannel=True)
     """
+    if multichannel:
+        if image.shape[2] not in (3, 4):
+            msg = "Input image must be grayscale, RGB, or RGBA; but has " \
```
Comment (stefanv) [inline skimage/restoration/_denoise.py]: remove backslash

File: skimage/restoration/_denoise.py [hunk 8]
```diff
@@ -62,8 +66,21 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
...
+                  "shape {0}."
+            warnings.warn(msg.format(image.shape))
+    else:
+        if image.ndim > 2:
+            msg = "Input image must be grayscale, RGB, or RGBA; but has " \
```
Comment (stefanv) [inline skimage/restoration/_denoise.py]: remove backslash

**Comment by stefanv** [comment]
One more comment!  Sorry, I missed those in my previous review.

File: skimage/restoration/_denoise.py [hunk 9]
```diff
@@ -62,8 +66,19 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
     >>> astro = astro[220:300, 220:320]
     >>> noisy = astro + 0.6 * astro.std() * np.random.random(astro.shape)
     >>> noisy = np.clip(noisy, 0, 1)
-    >>> denoised = denoise_bilateral(noisy, sigma_range=0.05, sigma_spatial=15)
+    >>> denoised = denoise_bilateral(noisy, sigma_range=0.05,
+                                     sigma_spatial=15, multichannel=True)
```
Comment (OrkoHunter) [inline skimage/restoration/_denoise.py]: I don't know how to deal with this. The [doctests](https://travis-ci.org/scikit-image/scikit-image/jobs/104567193#L2938) fail here. I really don't want to put 100 columns in a line.

Comment (jni) [inline skimage/restoration/_denoise.py]: @OrkoHunter 

```python
>>> denoised = denoise_bilateral(noisy, sigma_range=0.05,
...                              sigma_spatial=15, multichannel=True)
```
(Notice the `...`)

=)

Comment (OrkoHunter) [inline skimage/restoration/_denoise.py]: Awesome!

**Comment by jni** [comment]
More comments, sorry!

(1) the new lines need to be tested using `expected_warnings` and `assert_raises`, and
(2) the check for image shape in the first clause is a bit dangerous, `if img.shape[2] not in [3, 4]`: what if the input image is 2D grayscale? Thus, add a check for multichannel=True *and* img.ndim != 3, which I think should raise a ValueError.
(3) the warning for multichannel images with img.shape[2] > 4 should be slightly more informative and explain that the last axis is interpreted as channels. In particular, say that denoise_bilateral is implemented for *2D* grayscale or color images.

**Comment by OrkoHunter** [comment]
Hello @jni, I believe I addressed the last three points you mentioned in your last comment. If in case there's anything I left, please feel free to mention.

@stefanv Sorry for the delay ! Got little busy this week. I'll try to work on more issues this weekend.

File: doc/examples/filters/plot_denoise.py [hunk 10]
```diff
@@ -49,14 +49,14 @@
 ax[0, 1].imshow(denoise_tv_chambolle(noisy, weight=0.1, multichannel=True))
 ax[0, 1].axis('off')
 ax[0, 1].set_title('TV')
-ax[0, 2].imshow(denoise_bilateral(noisy, sigma_range=0.05, sigma_spatial=15))
+ax[0, 2].imshow(denoise_bilateral(noisy, sigma_range=0.05, sigma_spatial=15, multichannel=True))
```
Comment (ahojnnes) [inline doc/examples/filters/plot_denoise.py]: Wrap at 80.

Comment (OrkoHunter) [inline doc/examples/filters/plot_denoise.py]: Done !

File: skimage/restoration/_denoise.py [hunk 11]
```diff
@@ -62,8 +66,30 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
...
+                              "channels in last axis. ``denoise_bilateral`` is "
+                              "implemented for 2D grayscale and color images "
+                              "only.".format(image.shape, image.shape[2]))
+            else:
+                msg = "Input image must be grayscale, RGB, or RGBA; but has shape {0}."
+                warnings.warn(msg.format(image.shape))
+    else:
+        if image.ndim > 2:
+            msg = "Input image must be grayscale, RGB, or RGBA; but has shape {0}."
+            raise TypeError(msg.format(image.shape))
```
Comment (emmanuelle) [inline skimage/restoration/_denoise.py]: Shouldn't the error message mention the ``multichannel`` kw argument?

File: skimage/restoration/_denoise.py [hunk 12]
```diff
@@ -62,8 +66,30 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
...
+    >>> denoised = denoise_bilateral(noisy, sigma_range=0.05,
+    ...                              sigma_spatial=15, multichannel=True)
     """
+    if multichannel:
+        if image.ndim != 3:
+            raise ValueError("Use ``multichannel=False`` for 2D grayscale images. "
+                             "The last axis of the input image must be multiple "
+                             "color channels not another spatial dimension.")
+        elif image.shape[2] not in (3, 4):
+            if image.shape[2] > 4:
```
Comment (emmanuelle) [inline skimage/restoration/_denoise.py]: in principle, it's not a problem to have a lot of channels, eg for hyperspectral images. The principle of ``multichannel`` is to transfer to the user the responsibility of whether the last dimension should be considered as channels or not.

File: skimage/restoration/_denoise.py [hunk 13]
```diff
@@ -3,10 +3,11 @@
 from .. import img_as_float
 from ..restoration._denoise_cy import _denoise_bilateral, _denoise_tv_bregman
 from .._shared.utils import _mode_deprecations
+import warnings
 
 
 def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
-                      bins=10000, mode='constant', cval=0):
+                      bins=10000, mode='constant', cval=0, multichannel=False):
     """Denoise image using bilateral filter.
```
Comment (emmanuelle) [inline skimage/restoration/_denoise.py]: I think the default value of ``multichannel`` should be ``True``, since the case of 2-D images is more frequent than 3-D images.

File: skimage/restoration/_denoise.py (hunk 11 above)
Comment (emmanuelle) [inline skimage/restoration/_denoise.py]: Can we say "Bilateral filter is not implemented for grayscale images of 3 or more dimensions. Use ``multichannel=True`` for 2-D RGB images" ?

File: skimage/restoration/_denoise.py (hunk 13 above)
Comment (OrkoHunter) [inline skimage/restoration/_denoise.py]: That would be better.

File: skimage/restoration/_denoise.py (hunk 11 above)
Comment (OrkoHunter) [inline skimage/restoration/_denoise.py]: Sure, done.

Comment (OrkoHunter) [inline skimage/restoration/_denoise.py]: Sorry, the change broke some tests. I'll fix them soon !

**Comment by OrkoHunter** [comment]
Is this okay to go now?

**Comment by soupault** [comment]
@OrkoHunter don't think so :smiling_imp: - tests are failing.

**Comment by OrkoHunter** [comment]
Ah! Yes, I fogot to tweak the tests after changing the default value of `multichannel`.

**Comment by OrkoHunter** [comment]
Travis looks happy now =)

File: skimage/restoration/_denoise.py [hunk 14]
```diff
@@ -64,6 +68,30 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
...
+                              "as channels. Input image with shape {0} has {1} "
+                              "channels in last axis. ``denoise_bilateral`` is "
+                              "implemented for 2D grayscale and color images "
+                              "only.".format(image.shape, image.shape[2]))
+            else:
+                msg = "Input image must be grayscale, RGB, or RGBA; but has shape {0}."
+                warnings.warn(msg.format(image.shape))
+    else:
+        if image.ndim > 2:
+            raise TypeError("Bilateral filter is not implemented for "
```
Comment (jni) [inline skimage/restoration/_denoise.py]: I think this actually should be a value error, as above

File: skimage/restoration/_denoise.py [hunk 15]
```diff
@@ -64,6 +68,30 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
     >>> noisy = np.clip(noisy, 0, 1)
     >>> denoised = denoise_bilateral(noisy, sigma_range=0.05, sigma_spatial=15)
     """
+    if multichannel:
+        if image.ndim != 3:
```
Comment (jni) [inline skimage/restoration/_denoise.py]: This will actually be a confusing message if the input image is e.g. 4D, 3D + channels. So I recommend this message for image.ndim == 2, and a different message for image.ndim > 3.

Comment (OrkoHunter) [inline skimage/restoration/_denoise.py]: What should the error raised for `image.ndim > 3` ? Do we have to say that the function has only been implemented for 2D images?

Comment (jni) [inline skimage/restoration/_denoise.py]: yeah I would say it's only implemented for 2D grayscale (image.ndim == 2) and 2D multichannel (image.ndim == 3) images. Something along those lines.

Comment (OrkoHunter) [inline skimage/restoration/_denoise.py]: @jni Done. I also added a test for it. Thanks.

**Comment by jni** [comment]
:+1: from me

**Comment by soupault** [comment]
We really should think about something like this https://github.com/scikit-learn/scikit-learn/blob/a228016585c7f8093c4588ef40cea7e4c9c9daf2/sklearn/utils/validation.py#L271 .
I mean, outside this PR :).
Merging, thanks!

**Comment by OrkoHunter** [comment]
Thanks @soupault
